ompi/mca/coll/base/coll_base_scatter.c

```diff
@@ -273,5 +274,106 @@ ompi_coll_base_scatter_intra_basic_linear(const void *sbuf, int scount,
     return MPI_SUCCESS;
 }
 
-
 /* copied function (with appropriate renaming) ends here */
+
+int
+ompi_coll_base_scatter_intra_linear_nb(const void *sbuf, int scount,
+                                       struct ompi_datatype_t *sdtype,
+                                       void *rbuf, int rcount,
+                                       struct ompi_datatype_t *rdtype,
+                                       int root,
+                                       struct ompi_communicator_t *comm,
+                                       mca_coll_base_module_t *module,
+                                       int max_reqs)
+{
+    int i, rank, size, err, line, nreqs;
+    ptrdiff_t incr;
+    char *ptmp;
+    ompi_request_t **reqs = NULL, **preq;
+
+    rank = ompi_comm_rank(comm);
+    size = ompi_comm_size(comm);
+
+    /* If not root, receive data. */
+    if (rank != root) {
+        err = MCA_PML_CALL(recv(rbuf, rcount, rdtype, root,
+                                MCA_COLL_BASE_TAG_SCATTER,
+                                comm, MPI_STATUS_IGNORE));
+        if (MPI_SUCCESS != err) {
+            line = __LINE__; goto err_hndl;
+        }
+
+        return MPI_SUCCESS;
+    }
+
+    if (max_reqs <= 1) {
+        max_reqs = 0;
+        nreqs = size - 1; /* no send for myself */
+    } else {
+        nreqs = size - (size / max_reqs); /* avoid allocating too many reqs */
+    }
+
+    reqs = ompi_coll_base_comm_get_reqs(module->base_data, nreqs);
+    if (NULL == reqs) {
+        err = OMPI_ERR_OUT_OF_RESOURCE;
+        line = __LINE__; goto err_hndl;
+    }
+
+    err = ompi_datatype_type_extent(sdtype, &incr);
+    if (OMPI_SUCCESS != err) {
+        line = __LINE__; goto err_hndl;
+    }
+    incr *= scount;
+
+    /* I am the root, loop sending data. */
+    for (i = 0, ptmp = (char *)sbuf, preq = reqs; i < size; ++i, ptmp += incr) {
+        /* simple optimization */
+        if (i == rank) {
+            if (MPI_IN_PLACE != rbuf) {
+                err = ompi_datatype_sndrcv(ptmp, scount, sdtype, rbuf, rcount,
+                                           rdtype);
+            }
+        } else {
+            if (!max_reqs || (i % max_reqs)) {
+                err = MCA_PML_CALL(isend(ptmp, scount, sdtype, i,
```

Comment:
seems like we actually can have much more than max_reqs in flight..

Reply:
Good catch, the code writes outside the bounds of preq (the array of requests). This code works only when coll_tuned_scatter_algorithm_max_requests is less or equal to 1.

Reply:
@yosefe , what do you mean? Yes, we can, we just invoke blocking send every `max_reqs` 

@bosilca, in what case does it write outside the bounds? We allocate requests for every rank, except 'blocking send' cases, which is `size/max_reqs`.   
 It works with up to `size` max_reqs (was tested up to 256 max_reqs, 2048 ranks)

Reply:
let's assume max_reqs is 20 on a larger communicator, enough to make nreqs < (size - 1) - 1. Due to the usage of the % in the if on line 337,  all requests over max_reqs will  be sent as non-blocking and will be stored on preq (after it has been ++ by more than max_reqs). So you are writing outside the bounds you were granted on the preq array. It might happen that it  seems to be working simply because internally the request management allocates more than you requested, so you might not write outside the internally allocated array (only outside the range you were provided).

Reply:
Please check request allocation:
```
    if (max_reqs <= 1) {
        max_reqs = 0;
        nreqs = size - 1; /* no send for myself */
    } else {
        nreqs = size - (size / max_reqs); /* avoid allocating too many reqs */
    }
```

Reply:
Should I rename mca `coll_tuned_scatter_algorithm_max_requests` to `coll_tuned_scatter_algorithm_blocking_send_ratio`?

Reply:
Oh my, I now understand the real meaning of ```nreqs = size - (size / max_reqs)```, allocate space for all non-blocking requests and skip the one for the blocking one.

Reply:
sorry for the confusion, I just tried to re-use some mca name used by other collectives.
Do you tnink better rename it?

Reply:
names are fine but I would add a comment explaining the nreqs usage.